From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Fri, 31 Oct 2025 21:11:28 +0800
Subject: [PATCH] Add back read-only datapack command


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index e05f75b6bbd48d6fb7379926d68d5f8ed6014be9..11697e65a7f7871920ca7bad37de07f78a3da1f8 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -192,7 +192,7 @@ public class Commands {
         if(me.earthme.luminol.config.modules.experiment.CommandConfig.data) {
             DataCommands.register(this.dispatcher); // Folia - region threading - TODO
         }
-        //DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO
+        DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO // Luminol - Add back read-only datapack command
         //DebugCommand.register(this.dispatcher); // Folia - region threading - TODO
         DefaultGameModeCommands.register(this.dispatcher);
         //DialogCommand.register(this.dispatcher, context); // Folia - region threading - TODO
diff --git a/net/minecraft/server/commands/DataPackCommand.java b/net/minecraft/server/commands/DataPackCommand.java
index 3eaa9c85d06f6ecc8ed639d8befa861439390eb7..a5032b34752061d61fd9ddb4db828d35e403f5e7 100644
--- a/net/minecraft/server/commands/DataPackCommand.java
+++ b/net/minecraft/server/commands/DataPackCommand.java
@@ -98,7 +98,7 @@ public class DataPackCommand {
         dispatcher.register(
             Commands.literal("datapack")
                 .requires(Commands.hasPermission(2))
-                .then(
+                /*.then( // Luminol - Add back read-only datapack command
                     Commands.literal("enable")
                         .then(
                             Commands.argument("name", StringArgumentType.string())
@@ -160,11 +160,11 @@ public class DataPackCommand {
                                 .suggests(SELECTED_PACKS)
                                 .executes(commandContext -> disablePack(commandContext.getSource(), getPack(commandContext, "name", false)))
                         )
-                )
+                )*/ // Luminol - Add back read-only datapack command
                 .then(
                     Commands.literal("list")
                         .executes(commandContext -> listPacks(commandContext.getSource()))
-                        .then(Commands.literal("available").executes(commandContext -> listAvailablePacks(commandContext.getSource())))
+                        // .then(Commands.literal("available").executes(commandContext -> listAvailablePacks(commandContext.getSource()))) // Luminol - Add back read-only datapack command
                         .then(Commands.literal("enabled").executes(commandContext -> listEnabledPacks(commandContext.getSource())))
                 )
                 .then(
@@ -252,7 +252,7 @@ public class DataPackCommand {
     }
 
     private static int listPacks(CommandSourceStack source) {
-        return listEnabledPacks(source) + listAvailablePacks(source);
+        return listEnabledPacks(source) ;// + listAvailablePacks(source); // Luminol - Add back read-only datapack command
     }
 
     private static int listAvailablePacks(CommandSourceStack source) {
@@ -280,7 +280,7 @@ public class DataPackCommand {
 
     private static int listEnabledPacks(CommandSourceStack source) {
         PackRepository packRepository = source.getServer().getPackRepository();
-        packRepository.reload();
+        // packRepository.reload(); // Luminol - Add back read-only datapack command
         Collection<? extends Pack> selectedPacks = packRepository.getSelectedPacks();
         if (selectedPacks.isEmpty()) {
             source.sendSuccess(() -> Component.translatable("commands.datapack.list.enabled.none"), false);
