From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Tue, 28 Oct 2025 23:19:33 +0800
Subject: [PATCH] Fix possible delayed POI updates while portaling


diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index 6435d54ad704252f0f43e52735d86dca9b9df6b0..158c1f5d965bc5951db56c6b1ad99f0d8ea71571 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -540,6 +540,9 @@ public final class RegionizedWorldData {
     public final Map<Long, ChunkSectionInventoryEntityTracker> containerEntityMovementTrackerMap = new it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap<>();
     public final Map<Long, ChunkSectionItemEntityMovementTracker> itemEntityMovementTrackerMap = new it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap<>();
     // Luminol end
+    // Luminol start - Nether portal searach fixing
+    public boolean instantPOIUpdating = false;
+    // Luminol end
 
     public RegionizedWorldData(final ServerLevel world) {
         this.world = world;
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index eb3aef1f5198592eda5dafe84b91271e861b4135..985ddea9d74a37f58f32a27375f6f7e6af0f2192 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -2394,11 +2394,12 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         Optional<Holder<PoiType>> optional1 = PoiTypes.forState(newState);
         if (!Objects.equals(optional, optional1)) {
             BlockPos blockPos = pos.immutable();
-            optional.ifPresent(holder -> io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4, () -> { // Folia - region threading
+            final java.util.concurrent.Executor scheduler =  me.earthme.luminol.config.modules.fixes.PortalSearchConfig.instantPoiUpdatingForNetherPortal && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this, pos.getX() >> 4, pos.getZ() >> 4) && this.getCurrentWorldData().instantPOIUpdating ? Runnable::run : task -> io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4, task); // Luminol - Fix nether portal fix // Luminol - Portal serach fixes
+            optional.ifPresent(holder -> /*io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4,*/scheduler.execute( () -> { // Folia - region threading // Luminol - Portal serach fixes
                 this.getPoiManager().remove(blockPos);
                 DebugPackets.sendPoiRemovedPacket(this, blockPos);
             }));
-            optional1.ifPresent(holder -> io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4, () -> { // Folia - region threading
+            optional1.ifPresent(holder -> /*io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(this, blockPos.getX() >> 4, blockPos.getZ() >> 4,*/scheduler.execute( () -> { // Folia - region threading // Luminol - Portal serach fixes
                 // Paper start - Remove stale POIs
                 if (optional.isEmpty() && this.getPoiManager().exists(blockPos, ignored -> true)) {
                     this.getPoiManager().remove(blockPos);
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 34d802060e096c5acfbd9a66fab96c0e8c792fe1..f5539ca5964abb18124522a8aa7ce940517b2011 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4718,9 +4718,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                                 }
 
                                 // we do not have the correct entity reference here
+                                destination.getCurrentWorldData().instantPOIUpdating = true; // Luminol - Portal serach fixes
                                 BlockUtil.FoundRectangle createdPortal =
                                     destination.getPortalForcer().createPortal(targetPos, originalPortalDirection, null, portalCreateRadius).orElse(null);
                                 // if it wasn't created, passing null is expected here
+                                destination.getCurrentWorldData().instantPOIUpdating = false; // Luminol - Portal serach fixes
                                 portalFound.complete(createdPortal);
                             }
                         );
